verification/spec/connection-handshake/chm.tla

```diff
@@ -0,0 +1,241 @@
+      Stores the connection end for the remote chain.
+ ***************************************************************************)
+Connections ==
+    [
+        localEnd : ConnectionEnds,
+        remoteEnd : ConnectionEnds
+    ]
+
+
+(***************************** ConnectionHandshakeModules ******************
+    A set of records defining the CHM.
+    A CHM record contains the following fields:
+
+    - connectionState -- a string
+      Stores the current state of this connection. It has one of the
+      following values: "UNINIT", "INIT", "TRYOPEN", "OPEN".
+
+    - connection -- a connection with two ends
+      Stores the connection.
+
+    - chainHeight -- a height
+      Stores the height of the chain running alongside the CHM.
+ ***************************************************************************)
+ConnectionHandshakeModules ==
+    [
+        connectionState : ConnectionStates,
+        connection : Connections,
+        chainHeight : Heights
+        \* light client, consensus state, etc.. 
+    ]
+
+
+
+(******************************** Messages ********************************
+These messages are connection handshake specific.
+ ***************************************************************************)
+ConnectionHandshakeMessages ==
+    [type : {"ConnOpenInit"}, 
+     connection : Connections]
+\*    \union
+\*    [type : {"ConnOpenTry"},
+\*        connection : Connections,
+\*        stateProof : Proofs, 
+\*        consensusHeight : Proofs]
+
+
+
+(***************************************************************************
+ Connection handshake message handlers.
+***************************************************************************)
+
+\* The chModule handles a "ConnOpenInit" message.
+connectionOpenInit(chm, initMsg) ==
+  \* If we're in the right state.
+  IF chm.connectionState = "UNINIT" THEN
+         [initCHM |-> [chm EXCEPT !.connection = initMsg.connection,
+                                  !.connectionState = "INIT"],
+          msg |-> [type |-> "ConnOpenTry",
+                   connection |-> initMsg.connection]]
+                   \* TODO: Proofs go here.                   
+   \* Otherwise, do not update the chain
+   ELSE [initCHM |-> chm,
+         msg |-> noMsg ]
+
+handleMsgConnectionOpenInit ==
+ /\ inMsg.type = "ConnOpenInit"
+ /\ LET res == connectionOpenInit(chModule, inMsg) IN
+      /\ chModule' = res.initCHM
+      /\ outMsg' = res.msg
+      /\ UNCHANGED <<inMsg>>
+
+
+
+(***************************************************************************
+ Connection Handshake Module actions.
+ ***************************************************************************)
+
+\* Advance the height of the chain if MaxHeight is not yet reached.
+advanceChainHeight ==
+    /\ chModule.chainHeight < MaxHeight
+    /\ chModule' = [chModule EXCEPT
+                    !.chainHeight = chModule.chainHeight + 1
+                 ]
+    /\ UNCHANGED <<outMsg, inMsg>>
+
+
```

Comment:
I think we should clearly separate the environment and the ch spec, including the variables and functionality. Right now they are mixed up. If a handler accepts a msg and changes state it should cause `advanceChainHeight`. We should keep track of connection info per height.  We will also need to have the environment send the client state updates and the handler to simulate proof verification. For this we need to keep track of client information per height.

Reply:
Looking at fastsync.tla (tendermint-rs/blob/master/docs/spec/fastsync/fastsync.tla) the approach there is to have the network/environment/node and peers all bundled in a single .tla module.

In Ilina's relayer work (#55), in contrast, the environment is nicely separated: The relayer instantiates the environment and defines the the top-level spec.

For the sake of modularity, I agree we should have this separation between environment and CH spec. However, unlike the relayer spec, I think in our case the order will be reversed: the Environment would instantiate the CHM and would define the top-level spec.

Reply:
yeah, same for fastsync.tla. But better to separate. And at least in the scheduler.tla the vars for scheduler and env are clearly separated. Sorry i can't remember how it's done in the fastsync.tla spec.